Create RFID Server

neelp03 opened the pull request:
The RFID server has 3 endpoints: 

1. `GET /getRFIDs` - gets all RFID users' info
2. `POST /createRFID` - toggles between validating and creating
3. `DELETE /deleteRFID` - deletes an RFID user

For validating cards, we used pub/sub (AWS-IoT). ESP32 will publish to a topic that Core-v4 subscribes to. The validation results are sent back to ESP32 similarly. For creating new RFID cards, the RFID create API will toggle from validating to creating. So now when ESP32 sends the RFID data, it will create a new RFID user.

- [x]  Create APIs to get and delete RFID users
- [x]  Create AWS-IOT folder with info
- [x]  Create pub/sub to create and validate RFID cards
- [x]  Update RFID Schema
- [x] Add API tests

evanugarte commented on api/peripheral_api/routes/RFID.js:
```diff
@@ -0,0 +1,131 @@
+        message
+      } = JSON.parse(payload.toString());
+      const hash = await hashedByte(message);
+      RFID.findOne(
```
i thought we were iterating over every stored byte and comparing?

neelp03 replied:
Completely forgot about that. I will change it to that

evanugarte commented on package.json:
```diff
@@ -35,6 +35,7 @@
     "@fortawesome/react-fontawesome": "^0.1.14",
     "@tinymce/tinymce-react": "~3.8.4",
     "availity-reactstrap-validation": "^2.3.0",
+    "aws-iot-device-sdk": "^2.2.11",
```
i thought we only needed it for `api/`? to undo this do

```
npm uninstall aws-iot-device-sdk --save
```

neelp03 replied:
Okay so after uninstalling from `Core-v4/package.json` api-testing doesn't work, so I think we need it in both

evanugarte replied:
sounds good, resolving

evanugarte commented on test/RFID.js:
```diff
@@ -0,0 +1,156 @@
+const expect = chai.expect;
+const tools = require('./util/tools/tools');
+const { RfidHelper } = require('../api/peripheral_api/util/RFID-helpers');
+// let rfidHelper = sinon.createStubInstance(RfidHelper);
```
read through this file, if u see any random commented out code pls remove

evanugarte commented on api/config/config.example.json:
```diff
@@ -43,6 +43,8 @@
     "AccountID": "XXXXXXXXXX",
     "QueueName": "XXXXXXXXXX"
   },
+  "RFID_BCRYPT_SALT" : "NOT_SET",
```
the salt we use to hash passwords is `secretKey` on line 49,, can we just use the same salt?

evanugarte commented on api/config/AWS-IOT/README.md:
```diff
@@ -0,0 +1,6 @@
+PLEASE ADD THE FOLLOWING FILES TO THIS FOLDER TO CONNECT TO AWS IOT AND TO SUCCESSFULLY RUN THE RFID SERVER: 
+
+- AmazonRootCA1.pem
+- cert.pem.crt
+- private.pem.key
+- public.pem.key
```
lets add this to the [Getting Started](https://github.com/SCE-Development/Core-v4/wiki/Getting-Started) section on the Core-v4 wiki instead

evanugarte commented on api/peripheral_api/routes/RFID.js:
```diff
@@ -0,0 +1,79 @@
+
+  device
+    .on('connect', function() {
+      /* eslint-disable-next-line */
+      console.log('Connected to AWS IoT!');
+      device.subscribe('MessageForNode');
```
tbh i would remove, we can add logging to the sce backend differently
```suggestion
      device.subscribe('MessageForNode');
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,122 @@
+const fs = require('fs');
+
+
+class RfidHelper {
```
```suggestion
/**
 * Abstracts RFID card backend logic to a class to improve readability.
 * @member {bool} addRfid - Keeps track of the state of the server, whether
 * card data from the ESP32 should be verified or added to the database.
 * @member {string|null} name - The name of the user we should add to
 * MongoDB along with card data. When the server is in the add RFID state,
 * this variable is populated with said name.
 */
class RfidHelper {
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,122 @@
+    this.name = null;
+  }
+
+  addingRfid() {
```
```suggestion
  /**
   * Helper function to determine the state of the RFID API.
   * @returns {bool} If we are adding a card the value is true,
   * else false for verifying.
   */
  addingRfid() {
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,122 @@
+    return this.addRfid;
+  }
+
+  keysExist() {
```
```suggestion
  /**
   * Determines if the required files are present for the RFID server to work.
   * We check if an array of file names are present in api/config/AWS-IOT/
   * directory. A boolean value is returned depending on their presence.
   * @returns {bool} If the files are present the value is true,
   * else false.
   */
  keysExist() {
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,122 @@
+    return true;
+  }
+
+  hashedByte(byte) {
```
rename this to `hashCardData`, the associated docstring can be:
```js
  /**
   * Hashes the RFID card data to be stored in MongoDB.
   * @param {Number} byte The bytes of the RFID card formatted to an int.
   * For example if the bytes in hex are FF A1 B2 77, the param would become
   * "4288787063".
   * @returns {Promise<string|null>} Resolves with the hash of the byte data
   * or null if the bytes weren't able to be hashed.
   */
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,122 @@
+    });
+  }
+
+  getCards() {
```
see my above comments, think you can write a function comment yourself? hint: the return value for this would be `Promise<Array<Object>>`.

evanugarte commented on api/util/SceHttpServer.js:
```diff
@@ -85,7 +85,7 @@ class SceHttpServer {
         useCreateIndex: true,
       })
-      .then(() => {})
+      .then(() => { })
```
```suggestion
      .then(() => {})
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,122 @@
+    return new Promise(async (resolve, reject) => {
+      const newRFID = new RFID({
+        name: name,
+        byte: await this.hashedByte(JSON.parse(payload.toString()).message)
```
`hashedByte` has the potential to return `null`. shouldn't we account for this and `resolve(BAD_REQUEST)` if this happens?

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,122 @@
+  async validate(payload) {
+    return new Promise(async (resolve, reject) => {
+      const { message } = JSON.parse(payload.toString());
+      const hash = await this.hashedByte(message);
```
if `hashedByte` returns `null`, we should return early and `resolve(NOT_FOUND);`.

neelp03 commented:
I added function documentation for the RFID helper functions and removed all the unnecessary comments. As for the readme.MD file of AWS-IOT dir, I will add that to the Getting Started Page later tonight

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,195 @@
+  }
+
+  /**
+   * @param {Wrapper} device  MQTT Client Wrapper
+   * @param {JSONObject} payload Incoming card data from ESP32
+   * Handles incoming messages from ESP32
+   * If in creating mode, then it will create new RFID card
+   * else validate RFID card
```
```suggestion
   * Handles incoming messages from ESP32
   * If in creating mode, then it will create new RFID card
   * else validate RFID card
   * @param {Wrapper} device  MQTT Client Wrapper
   * @param {JSONObject} payload Incoming card data from ESP32
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,195 @@
+   * @param {String|null} byte RFID card data
+   * @returns a hashed value if successful else null
+   */
+  hashedCardData(byte) {
```
```suggestion
  hashCardData(byte) {
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,195 @@
+  async validate(payload) {
+    return new Promise(async (resolve, reject) => {
+      const { message } = JSON.parse(payload.toString());
+      const hash = await this.hashedCardData(message);
```
```suggestion
      const hash = await this.hashCardData(message);
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,195 @@
+   * @returns {Number} 404 if card data was null or not in the database
+   * else 200 if the incoming data is not null and it exists in the database
+   */
+  async validate(payload) {
```
little more descriptive
```suggestion
  async validateCard(payload) {
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,195 @@
+    return new Promise(async (resolve, reject) => {
+      const newRFID = new RFID({
+        name: name,
+        byte: await this.hashedCardData(JSON.parse(payload.toString()).message)
```
```suggestion
        byte: await this.hashCardData(JSON.parse(payload.toString()).message)
```

evanugarte commented on api/peripheral_api/util/RFID-helpers.js:
```diff
@@ -0,0 +1,195 @@
+      this.addRfid = false;
+      clearTimeout();
+    } else {
+      const validateResponse = await this.validate(payload);
```
```suggestion
      const validateResponse = await this.validateCard(payload);
```

evanugarte reviewed (approved):
Looks good, open an issue saying "Add tests to util/RFID-helpers.js". We can do it another time, it will be good practice.

WillN-byte commented:
Everything looks great.
